Danny Ryan

Lab 12 

Dimsenionality Reduction 

3/29/2020

In [ ]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns 

In [ ]:
file_link = "https://drive.google.com/uc?export=download&id=1vqMBid4r0C8apwYcpBQfVQ_TwEUNWY8r"
df = pd.read_csv(file_link)
df.head()

0         1         2         3  ...        27        28        29   30
0  2.536741 -3.301172 -0.401756 -1.202885  ...  2.782064  1.896763 -0.039366  0.0
1  2.700367 -2.672759 -0.683936 -2.410570  ...  2.419307  1.033950 -0.033414  0.0
2 -1.020962 -0.851870  1.270827 -0.921015  ...  1.196119  1.751840  1.279966  1.0
3 -1.450912 -1.156441 -2.247361  1.744520  ...  1.796661  0.033870  0.763104  0.0
4  2.239896  2.216549 -2.511707  0.581304  ... -2.624473 -4.741839 -0.050563  0.0

[5 rows x 31 columns]

In [ ]:
df.describe

<bound method NDFrame.describe of             0         1         2         3  ...        27        28        29   30
0    2.536741 -3.301172 -0.401756 -1.202885  ...  2.782064  1.896763 -0.039366  0.0
1    2.700367 -2.672759 -0.683936 -2.410570  ...  2.419307  1.033950 -0.033414  0.0
2   -1.020962 -0.851870  1.270827 -0.921015  ...  1.196119  1.751840  1.279966  1.0
3   -1.450912 -1.156441 -2.247361  1.744520  ...  1.796661  0.033870  0.763104  0.0
4    2.239896  2.216549 -2.511707  0.581304  ... -2.624473 -4.741839 -0.050563  0.0
..        ...       ...       ...       ...  ...       ...       ...       ...  ...
995 -0.650636 -1.413686  0.867949 -0.407801  ...  0.918315  2.306975  1.148003  0.0
996 -0.660438 -4.380001  1.160266  0.290124  ...  3.335666  6.682773 -1.730456  0.0
997 -1.117849  0.570688 -1.660963  1.255061  ... -1.000558 -0.628699  0.501853  0.0
998 -1.813400 -1.001267 -0.628922 -0.721671  ...  1.195701  1.206945 -0.280884  0.0
999 -1.992246 -0.262096 -0.416985 -1.41725

In [ ]:
X = df.iloc[:, : 30]

In [ ]:
X.describe

<bound method NDFrame.describe of             0         1         2  ...        27        28        29
0    2.536741 -3.301172 -0.401756  ...  2.782064  1.896763 -0.039366
1    2.700367 -2.672759 -0.683936  ...  2.419307  1.033950 -0.033414
2   -1.020962 -0.851870  1.270827  ...  1.196119  1.751840  1.279966
3   -1.450912 -1.156441 -2.247361  ...  1.796661  0.033870  0.763104
4    2.239896  2.216549 -2.511707  ... -2.624473 -4.741839 -0.050563
..        ...       ...       ...  ...       ...       ...       ...
995 -0.650636 -1.413686  0.867949  ...  0.918315  2.306975  1.148003
996 -0.660438 -4.380001  1.160266  ...  3.335666  6.682773 -1.730456
997 -1.117849  0.570688 -1.660963  ... -1.000558 -0.628699  0.501853
998 -1.813400 -1.001267 -0.628922  ...  1.195701  1.206945 -0.280884
999 -1.992246 -0.262096 -0.416985  ...  0.233993  0.861341  0.098580

[1000 rows x 30 columns]>

In [ ]:
y = df.iloc[:,30]
y.shape

(1000,)

In [ ]:
y.describe

<bound method NDFrame.describe of 0      0.0
1      0.0
2      1.0
3      0.0
4      0.0
      ... 
995    0.0
996    0.0
997    0.0
998    0.0
999    0.0
Name: 30, Length: 1000, dtype: float64>

In [ ]:
from sklearn.model_selection import train_test_split
Xtrain, Xtest, ytrain, ytest = train_test_split(X,y, test_size = 0.5, random_state = 0)

In [ ]:
from sklearn.decomposition import PCA

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

In [ ]:
pca = PCA(n_components=2)
pca.fit(Xtrain)
Xtrain2 = pca.transform(Xtrain)
Xtest2 = pca.transform(Xtest)

In [ ]:
Xtrain2.shape

(500, 2)

In [ ]:
pca.explained_variance_ratio_

array([0.35609967, 0.19444318])

In [ ]:
model = KNeighborsClassifier(n_neighbors=1)
model.fit(Xtrain, ytrain)
ypred = model.predict(Xtest)
print("accurecy" , accuracy_score(ytest, ypred))

accurecy 0.94


In [ ]:
model1 = KNeighborsClassifier(n_neighbors= 1)
model1.fit(Xtrain2, ytrain)
ypred1 = model1.predict(Xtest2)
print("accurecy", accuracy_score(ytest, ypred1))

accurecy 0.59


Using the orginal training data, the accuracy was 0.94, and after transforming the training data, the accuracy 0.59. This was not suprising, as the original data set had 30 columns to help differientaite the data, wheras the transformed data only had two columns, so it would inherintly have less variability, and thus make it tougher to predict. 

In [ ]:
a = np.arange(2,30,1)

for i in (a):
  pca = PCA(n_components=i)
  pca.fit(Xtrain)
  Xtrain2 = pca.transform(Xtrain)
  Xtest2 = pca.transform(Xtest)
  model1 = KNeighborsClassifier(n_neighbors= 1)
  model1.fit(Xtrain2, ytrain)
  ypred1 = model1.predict(Xtest2)
  print(i , "--", " accurecy ", accuracy_score(ytest, ypred1))
  sum = np.sum(pca.explained_variance_ratio_)
  print(i,"-- variance ", sum)
  print("=================")


2 --  accurecy  0.59
2 -- variance  0.5505428529752316
3 --  accurecy  0.892
3 -- variance  0.7000931034616613
4 --  accurecy  0.896
4 -- variance  0.8047525987162285
5 --  accurecy  0.958
5 -- variance  0.8678723134134508
6 --  accurecy  0.958
6 -- variance  0.8842434923969543
7 --  accurecy  0.962
7 -- variance  0.8995492052460975
8 --  accurecy  0.96
8 -- variance  0.9136161590815112
9 --  accurecy  0.956
9 -- variance  0.9273018337397133
10 --  accurecy  0.946
10 -- variance  0.9406715138279935
11 --  accurecy  0.95
11 -- variance  0.9535610392198978
12 --  accurecy  0.952
12 -- variance  0.965869254451283
13 --  accurecy  0.948
13 -- variance  0.9778824434371968
14 --  accurecy  0.946
14 -- variance  0.9892153892474123
15 --  accurecy  0.94
15 -- variance  1.0
16 --  accurecy  0.94
16 -- variance  0.9999999999999999
17 --  accurecy  0.94
17 -- variance  0.9999999999999999
18 --  accurecy  0.94
18 -- variance  0.9999999999999999
19 --  accurecy  0.94
19 -- variance  0.9999999999999

In [ ]:
ytrain.shape

(500,)

The best accurecy I could get was 0.962 using 7 components, and the percent of the original variance was 0.899.